Account for blocks in maybe_move_cursor_to_next_token

Turns out #747 was indeed a local config issue (lexical was compiled with a more recent elixir/erlang version than the project).

However, the error was hiding the true underlying issue:

```
2024-05-22T20:43:39.721 helix_lsp::transport [ERROR] lexical <- InternalError: ** (ErlangError) Erlang error: {:exception, {:badmatch, #Sourceror.Zipper<
    [{{:__block__, [format: :keyword, line: 848, column: 15], [:enforce]}, {:__block__, [line: 848, column: 24], [true]}}, {{:__block__, [format: :keyword, line: 848, column: 30], [:module]}, {:__aliases__, [last: [line: 848, column: 38], line: 848, column: 38], [:CreatedContract]}}]
    #...
  >}, [{LXical.RemoteControl.CodeIntelligence.Definition, :maybe_move_cursor_to_next_token, 4, [file: ~c"lib/lexical/remote_control/code_intelligence/definition.ex", line: 130]}, {LXical.RemoteControl.CodeIntelligence.Definition, :parse_location, 2, [file: ~c"lib/lexical/remote_control/code_intelligence/definition.ex", line: 111]}, {LXical.RemoteControl.CodeIntelligence.Definition, :definition, 2, []}]}
    (kernel 9.1) erpc.erl:702: :erpc.call/5
    (lx_server 0.5.0) lib/lexical/server/provider/handlers/go_to_definition.ex:9: LXical.Server.Provider.Handlers.GoToDefinition.handle/2
    (lx_server 0.5.0) lib/lexical/server/provider/queue.ex:99: anonymous fn/2 in LXical.Server.Provider.Queue.State.as_task/2
    (elixir 1.15.6) lib/task/supervised.ex:101: Task.Supervised.invoke_mfa/2
    (elixir 1.15.6) lib/task/supervised.ex:36: Task.Supervised.reply/4
```

--- a/apps/remote_control/lib/lexical/remote_control/code_intelligence/definition.ex
+++ b/apps/remote_control/lib/lexical/remote_control/code_intelligence/definition.ex
@@ -1,5 +1,5 @@
 defmodule Lexical.RemoteControl.CodeIntelligence.Definition do
-  alias Future.Code, as: Code
+  alias Future.Code
   alias Lexical.Ast
   alias Lexical.Ast.Analysis
   alias Lexical.Document
@@ -125,20 +125,18 @@ defmodule Lexical.RemoteControl.CodeIntelligence.Definition do
        when type in [:function, :module, :macro] do
     position = Position.new(document, line, column)
 
-    case Ast.zipper_at(document, position) do
-      {:ok, zipper} ->
-        zipper = %{node: {entity_name, meta, _}} = Sourceror.Zipper.next(zipper)
+    with {:ok, zipper} <- Ast.zipper_at(document, position),
+         %{node: {entity_name, meta, _}} <- Sourceror.Zipper.next(zipper) do
+      meta =
+        if entity_name == :when do
+          %{node: {_entity_name, meta, _}} = Sourceror.Zipper.next(zipper)
+          meta
+        else
+          meta
+        end
 
-        meta =
-          if entity_name == :when do
-            %{node: {_entity_name, meta, _}} = Sourceror.Zipper.next(zipper)
-            meta
-          else
-            meta
-          end
-
-        {meta[:line], meta[:column]}
-
+      {meta[:line], meta[:column]}
+    else
       _ ->
         {line, column}
     end

--- a/apps/remote_control/test/fixtures/navigations/lib/macro_struct.ex
+++ b/apps/remote_control/test/fixtures/navigations/lib/macro_struct.ex
@@ -1,0 +1,32 @@
+defmodule MacroStruct do
+  defmacro __using__(_) do
+    Module.register_attribute(__CALLER__.module, :field_names, accumulate: true)
+
+    quote do
+      import(unquote(__MODULE__), only: [field: 2, typedstruct: 2])
+    end
+  end
+
+  defmacro __before_compile__(_) do
+    fields = __CALLER__.module |> Module.get_attribute(:field_names) |> List.wrap()
+
+    quote do
+      defstruct unquote(fields)
+    end
+  end
+
+  defmacro typedstruct(opts, do: body) do
+    Module.put_attribute(__CALLER__.module, :opts, opts)
+
+    quote do
+      defmodule unquote(opts[:module]) do
+        @before_compile unquote(__MODULE__)
+        unquote(body)
+      end
+    end
+  end
+
+  defmacro field(name, _type) do
+    Module.put_attribute(__CALLER__.module, :field_names, name)
+  end
+end

--- a/apps/remote_control/test/fixtures/navigations/lib/struct.ex
+++ b/apps/remote_control/test/fixtures/navigations/lib/struct.ex
@@ -1,0 +1,11 @@
+defmodule NormalStruct do
+  defstruct [:variant]
+end
+
+defmodule TypedStructs do
+  use MacroStruct
+
+  typedstruct enforce: true, module: MacroBasedStruct do
+    field(:contract_id, String.t())
+  end
+end

--- a/apps/remote_control/test/lexical/remote_control/code_intelligence/definition_test.exs
+++ b/apps/remote_control/test/lexical/remote_control/code_intelligence/definition_test.exs
@@ -421,8 +421,7 @@ defmodule Lexical.RemoteControl.CodeIntelligence.DefinitionTest do
         end
       ]
 
-      {:ok, [location1, location2]} =
-        definition(project, subject_module, [uri, subject_uri])
+      {:ok, [location1, location2]} = definition(project, subject_module, [uri, subject_uri])
 
       {referenced_uri, definition_line} = location1
       assert definition_line =~ ~S[  def «greet(name)» do]
@@ -431,6 +430,21 @@ defmodule Lexical.RemoteControl.CodeIntelligence.DefinitionTest do
       {referenced_uri, definition_line} = location2
       assert definition_line == ~S[  defdelegate «greet(name)», to: MyDefinition]
       assert referenced_uri == subject_uri
+    end
+  end
+
+  describe "edge cases" do
+    setup [:with_referenced_file]
+
+    test "doesn't crash with structs defined with DSLs", %{project: project, uri: uri} do
+      subject_module = ~q[
+      defmodule MyTest do
+        def my_test(%TypedStructs.MacroBased|Struct{}) do
+        end
+      end
+      ]
+
+      assert {:ok, _file, _definition} = definition(project, subject_module, [uri])
     end
   end
 

